intake/source/discovery.py

```diff
@@ -12,19 +12,158 @@
+                     entrypoint.object_name)
+        if name in group:
+            shadowed = group[name]
+            logger.debug("Config shadowed entrypoint '%s = %s.%s'",
+                         shadowed.name,
+                         shadowed.module_name,
+                         shadowed.object_name)
+        if name in package_scan_results:
+            cls = package_scan_results[name]
+            del package_scan_results[name]
+            logger.debug("Config shadowed package scan result '%s = %s.%s'",
+                         name, cls.__module__, cls.__name__)
+        group[name] = entrypoint
 
-    This function searches path (or sys.path) for packages with names that
-    start with plugin_prefix.  Those modules will be imported and scanned for
-    subclasses of intake.source.base.Plugin.  Any subclasses found will be
-    instantiated and returned in a dictionary, with the plugin's name attribute
-    as the key.
+    # Discovery is complete.
+    
+    # Load entrypoints. Any that were shadowed or banned have already been
+    # removed above.
+    drivers = {}
+    for entrypoint in group.values():
+        try:
+            drivers[entrypoint.name] = _load_entrypoint(entrypoint)
+        except ConfigurationError:
+            logger.exception()
+            continue
+        logger.debug("Loaded entrypoint '%s = %s.%s'",
+                     entrypoint.name,
+                     entrypoint.module_name,
+                     entrypoint.object_name)
+
+    # TODO Un-comment this code to unleash warnings after most known intake_*
+    # packages have been updated to provide 'intake.drivers' entrypoints.
+    # if package_scan_results:
+    #     warnings.warn(
+    #         f"The drivers {package_scan_results} do not specify entry_points "
+    #         f"and were only discovered via a package scan. This may break in a "
+    #         f"future release of intake. The packages should be updated.")
+
+    # Now include any package scan results. Any that were shadowed or
+    # banned have already been removed above.
+    for name, cls in package_scan_results.items():
+        drivers[name] = cls
+        logger.debug("Loaded package scan result '%s = %s.%s'",
+                     name,
+                     cls.__module__,
+                     cls.__name__)
+
+    return drivers
+
+
+def _load_entrypoint(entrypoint):
+    """
+    Call entrypoint.load() and, if it fails, raise context-specific errors.
+    """
+    try:
+        return entrypoint.load()
+    except ModuleNotFoundError:
+        raise ConfigurationError(
+            f"Failed to load {entrypoint.name} driver because module "
+            f"{entrypoint.module_name} could not be found.")
+    except AttributeError:
+        raise ConfigurationError(
+            f"Failed to load {entrypoint.name} driver because no object "
+            f"named {entrypoint.object_name} could be found in the module "
+            f"{entrypoint.module_name}.")
```

Comment:
If py3+ only, you may want to `raise ... from e`, the ModuleNotFound may not be the entrypoint itself, but one of its dependencies.  Same for below. AttributeError, it may be loading the entrypoint trigger some unrelated AttributeErrors, in which case the error message is might be wrong. It is the most likely cause – I agree –, but not always and, not raising from may make the issue quite hard to track.

Reply:
+1 (or at least logging)